File: backend/domains/story/services/naturalLanguage/text.js
// ============================================================================
// Text helpers shared by every natural-language module
// ----------------------------------------------------------------------------
// Deliberately tiny and dependency-free: these are the only primitives the
// phrase modules are allowed to share, so a change here is easy to reason about.
// ============================================================================

/** Pick one element from an array using the supplied rng (seedable for tests). */
export function pick(arr, rng = Math.random) {
  return arr[Math.floor(rng() * arr.length)];
}

/** Join a list in prose: "a", "a and b", "a, b and c". Blanks are dropped. */
export function joinList(arr) {
  const xs = (arr || []).filter(Boolean);
  if (xs.length === 0) return '';
  if (xs.length === 1) return xs[0];
  if (xs.length === 2) return `${xs[0]} and ${xs[1]}`;
  return `${xs.slice(0, -1).join(', ')} and ${xs[xs.length - 1]}`;
}

/** Uppercase the first character, leaving the rest untouched. */
export function capitalize(text) {
  if (!text) return '';
  return text.charAt(0).toUpperCase() + text.slice(1);
}

/** Trimmed " — description" suffix, or '' when there is nothing to append. */
export function descriptionSuffix(description) {
  return description && description.trim() ? ` — ${description.trim()}` : '';
}

/** Turn a snake_case type into a readable label ("fortified_town" -> "fortified town"). */
export function readableType(type) {
  return type ? String(type).replace(/_/g, ' ') : '';
}

/** Region objects or bare strings -> plain region names. */
export function regionNamesOf(regions) {
  return (regions || []).map((r) => (typeof r === 'string' ? r : r.name));
}

File: story-engine/app/text.py
# ============================================================================
# Text helpers shared by every natural-language module
# ----------------------------------------------------------------------------
# Port of backend/domains/story/services/naturalLanguage/text.js. Deliberately
# tiny and dependency-free.
# ============================================================================
import random
import re


def pick(arr, rng=random.random):
    """Pick one element from a list using the supplied rng (seedable for tests)."""
    return arr[int(rng() * len(arr))]


def join_list(arr):
    """Join a list in prose: "a", "a and b", "a, b and c". Blanks are dropped."""
    xs = [x for x in (arr or []) if x]
    if len(xs) == 0:
        return ''
    if len(xs) == 1:
        return xs[0]
    if len(xs) == 2:
        return f'{xs[0]} and {xs[1]}'
    return f"{', '.join(xs[:-1])} and {xs[-1]}"


def capitalize(text):
    """Uppercase the first character, leaving the rest untouched."""
    if not text:
        return ''
    return text[0].upper() + text[1:]


def description_suffix(description):
    """Trimmed " — description" suffix, or '' when there is nothing to append."""
    if description and description.strip():
        return f' — {description.strip()}'
    return ''


def readable_type(type_):
    """Turn a snake_case type into a readable label ("fortified_town" -> "fortified town")."""
    return re.sub(r'_', ' ', type_) if type_ else ''


def region_names_of(regions):
    """Region objects or bare strings -> plain region names."""
    return [r if isinstance(r, str) else r.get('name') for r in (regions or [])]
